plugins/optimization-detective/site-health.php

```
@@ -0,0 +1,304 @@
+ *
+ * @param bool $in_plugin_row Whether the notice is to be printed in the plugin row.
+ */
+function od_maybe_render_rest_api_health_check_admin_notice( bool $in_plugin_row ): void {
+	if ( ! od_is_rest_api_unavailable() ) {
+		return;
+	}
+
+	$response = od_get_rest_api_health_check_response( true );
+	$result   = od_compose_site_health_result( $response );
+	if ( 'good' === $result['status'] ) {
+		// There's a slight chance the DB option is stale in the initial if statement.
+		return;
+	}
+
+	$message = sprintf(
+		$in_plugin_row
+			? '<summary style="margin: 0.5em 0">%s %s</summary>'
+			: '<p><strong>%s %s</strong></p>',
+		esc_html__( 'Warning:', 'optimization-detective' ),
+		esc_html( $result['label'] )
+	);
+
+	$message .= $result['description']; // This has already gone through Kses.
+
+	if ( current_user_can( 'view_site_health_checks' ) ) {
+		$site_health_message = wp_kses(
+			sprintf(
+				/* translators: %s is the URL to the Site Health admin screen */
+				__( 'Please visit <a href="%s">Site Health</a> to re-check this once you believe you have resolved the issue.', 'optimization-detective' ),
+				esc_url( admin_url( 'site-health.php' ) )
+			),
+			array( 'a' => array( 'href' => array() ) )
+		);
+		$message .= "<p><em>$site_health_message</em></p>";
+	}
+
+	if ( $in_plugin_row ) {
+		$message = "<details>$message</details>";
+	}
+
+	wp_admin_notice(
+		$message,
+		array(
+			'type'               => 'warning',
+			'additional_classes' => $in_plugin_row ? array( 'inline', 'notice-alt' ) : array(),
+			'paragraph_wrap'     => false,
+		)
+	);
+}
+
+/**
+ * Displays an admin notice on the plugin row if the REST API health check fails.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param string $plugin_file Plugin file.
+ */
+function od_render_rest_api_health_check_admin_notice_in_plugin_row( string $plugin_file ): void {
+	if ( 'optimization-detective/load.php' !== $plugin_file ) { // TODO: What if a different plugin slug is used?
```

I think we could use `plugin_basename( __FILE__ )` to address the `TODO` here?
```suggestion
	if ( plugin_basename( __FILE__ ) !== $plugin_file ) {
```

---

Looks like that doesn't work, at least not in the wp-env test environment. Here `plugin_basename( __FILE__ )`  is returning `performance/plugins/optimization-detective/site-health.php`.

---

Note that Embed Optimizer has similar code running at the `after_plugin_row_meta` filter:

https://github.com/WordPress/performance/blob/96bcb2d38b462e84ae7c200fb7d381358d2d4aeb/plugins/embed-optimizer/hooks.php#L345-L349

---

That should only be a concern in a "weird" development setup like `wp-env` though, for regular sites the plugin basename will be `optimization-detective/load.php`.

I wonder how WordPress computes the value for `$plugin_file` if it's not matching `plugin_basename( $plugin_root_dir_file )`. 🤔 

Anyway, what we could potentially do for now is create a function like `optimization_detective_get_basename()` that calls `plugin_basename( __FILE__ )` and uses it unless there is more than one `/` character in it, in which case it's a "weird setup" situation and we can fall back to the hard-coded one.

---

I guess we could do this:

```suggestion
	if ( basename( __DIR__ ) !== dirname( $plugin_file ) ) {
```

---

Maybe even slightly safer as:
```suggestion
	if ( basename( __DIR__ ) . '/load.php' !== $plugin_file ) {
```

---

Granted, the Embed Optimizer logic really depends on the installed plugin slugs actually being the same ones as on the directory. Since this is just the addition of a notice which will be showing up in Site Health anyway, maybe we should just accept that there's a slight chance a notice won't be shown given it would very unlikely for anything but the official slug to be used:

```suggestion
	if ( 'optimization-detective/load.php' !== $plugin_file ) {
```

---

I agree it's not critical, but I also think the hard-coding is not ideal. So maybe let's just keep the `TODO` in place for now and think about this separately later.